Provide top level method for rendering subtree (passes context)

jsfb opened the pull request:
Provide top level method for rendering subtree (passes context).  I almost wonder if we want to put an underscore in front of this new top level API, since it's maybe not something we actually support yet?

browniefed commented:
Is this going to replace withContext?

jsfb commented:
@browniefed To be determined.  We're still trying to figure out a good story around layers/portals.  Right now, this is our solution stop-gap solution to allow context to propagate to children, at least until we figure out the whole story.  This will probably be undocumented (like the rest of context) and subject to change.

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +457,13 @@ var ReactMount = {
    * @return {ReactComponent} Component instance rendered in `container`.
    */
-  render: function(nextElement, container, callback) {
+  renderSubtreeIntoContainer: function(parentComponent, nextElement, container, callback) {
```
I don't like the order of arguments here. If we're going to make this public we probably want some additional error messages (eg, if I call this method as a user without a `parentComponent` then we should hit an invariant).

If we make this an instance method it could have the same signature as `React.render`...

jsfb replied:
Yeah, I think Ben was also in favor of putting it on the component in order to preserve argument order.  I sort of like it too, because I see it like "asking this component to render a child" which seems better than rendering a child on behalf of the component.  But since I don't think this should be public API yet (I would have underscored it), I think addons is the right place for it, as per our discussion.  Moved it there.

zpao replied:
You didn't address the first part of this. What about warnings when called directly without a `parentComponent`?

jsfb commented:
This went up yesterday; any comments or ready to go?

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -524,11 +531,13 @@ var ReactMount = {
-      shouldReuseMarkup
+      shouldReuseMarkup,
+      parentComponent != null ?
+        parentComponent._reactInternalInstance._processChildContext(parentComponent._reactInternalInstance._context) :
```
Can you try to reformat and fit in 80

jsfb replied:
Done, reformatted.

zpao commented:
Are we going to expose this in `React.addons`?

jsfb commented:
Yes, my guess is that @sebmarkbage wants this exposed.  Done.

sebmarkbage commented:
@zpao's comment about throwing when no parent is provided.

zpao commented:
FWIW, I still think this name and order of arguments is pretty bad. We've never called anything a "subtree" let alone use that in a method name. Ditto with "intoContainer", `render` also renders into a container but we don't include that in the method name. So I sort of want to bikeshed a bit more before this gets checked in.

`renderWithParentContext(element, node, parent, callback)` or `renderWithParent`, `renderAsChild` … or something. Pretty much just anything without "subtree" and "container". And I want parent to be the 3rd argument, not the first.

Or even `renderWithContext(element, node, context, callback)` and skip the part where we actually pull the context off an component and require that to happen by the user (`React.addons.renderWithContext(<Layer/>, document.body, this.context)`)

(apparently this last idea isn't a good one - I talked with @jsfb and he said we might need to pull more than the context, also the thing about context being masked).

sebmarkbage commented:
Note that we know that this API is broken going in. It is just a workaround/hack until we can figure out the real one. This is why I wanted this to be a hack on the stack of life-cycles and just use `React.render` to avoid bikeshedding and confusion.

IMO, if we have to introduce another API it makes sense to call it subtree because it is a new concept that we're introducing.

This concept is not limited to context. There are many other things that affect subtrees that we might want to bring into this:

1) Debugging structure. E.g. tooltips shouldn't show up in the top level of React DevTools but rather in their respective subtrees.
2) Automatic clean up if the parent tree is unmounted.
3) Event bubbling.
4) Reconciliation order (`_mountOrder`).
5) Maybe refs unless we fully kill old-style refs? Maybe more?

zpao commented:
`renderSubtree(element, node, parent, callback)`?

jsfb commented:
That signature looks more "unexpected" to me, because it looks so similar but is actually subtly different.  I would glance at the function and think I knew the signature, only to be surprised by the parameter in the middle.

Also, it splits the parent from the child in an awkward way.

But I don't have a strong preference one way or the other.  If Paul likes this version better and there are no objections, I'll make the change.

Last call for objections @sebmarkbage or @spicyj.

sebmarkbage commented:
Meh. I thought @jsfb's original signature was intuitive. Seems arbitrary. Pick either one. This API probably can't easily support all the things we want to do with subtrees anyway so we'll likely update it later anyway, as we've discussed.

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +457,13 @@ var ReactMount = {
    * perform an update on it and only mutate the DOM as necessary to reflect the
    * latest React component.
    *
+   * @param {ReactElement} parentComponent The conceptual parent of this render tree.
```
Note: parent is not an element

jsfb commented:
Fixed to throw if the user passes in null or a non-component as the subtree parent.  (Also fixed jsdoc per @zpao).  I actually like the old behavior where if the first argument is null then it behaves exactly like React.render() - that feels very natural to me.  Let me know if the null behavior would be preferable (we'd still keep the check for a valid component though, to prevent mistakes - it would accept null or a valid component).

I opted to keep the function signature as-is.  Mainly because as I was changing it, I realized it would mean that most calls become renderSubtree(myElement, myContainer, this, callback) - and having this as the third parameter just felt so wrong.  @zpao Let me know if you'd like to overrule and I'll change the signature.

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +457,21 @@ var ReactMount = {
+      parentComponent != null && parentComponent._reactInternalInstance != null,
+      'parentComponent must be a valid React Component'
+    );
+  },
```
You forgot to call `_renderSubtreeIntoContainer` here

jsfb replied:
I'm a little surprised the unit test didn't blow up there.  It seems like travis should have failed this.  Interesting.

zpao commented:
My feelings still hold but I'm done fighting it. Let's get this done.

I want the null behavior to throw because we have an API for that and it's called `React.render`. `React.renderSubtreeIntoContainer` shouldn't be used there.

> most calls become renderSubtree(myElement, myContainer, this, callback)

Sounds like you want an instance method :P

jsfb commented:
Re null behavior: Ok, I can agree with that.

Haha, yeah, maybe I do want it to be an instance method, idk.  Whatever, we can figure out if it should be a top-level api or a component api if/when we ever move it out of addons.  Personally, my hope is that we'll replace the API with a proper layers solution and just delete this function altogether.

jsfb commented:
New diff added a couple unit tests, and added some functionality that @sebmarkbage wanted to prevent invocations from outside a lifecycle.

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +458,27 @@ var ReactMount = {
    */
-  render: function(nextElement, container, callback) {
+  renderSubtreeIntoContainer: function(parentComponent, nextElement, container, callback) {
+    console.log(typeof parentComponent._reactInternalInstance);
```
You'll want to get rid of this

zpao replied:
You missed this

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +458,27 @@ var ReactMount = {
+      ReactContext.inContextLifecycle,
+      'Must be in a lifecycle method to invoke renderSubtreeIntoContainer'
+    );
+    ReactMount._renderSubtreeIntoContainer(parentComponent, nextElement, container, callback);
```
`return` the result

zpao commented on src/browser/ui/ReactMount.js:
```diff
@@ -451,12 +458,27 @@ var ReactMount = {
+    );
+    invariant(
+      ReactContext.inContextLifecycle,
+      'Must be in a lifecycle method to invoke renderSubtreeIntoContainer'
```
What happens if I call this twice in row? Or in the callback? is the `inContextLifecyle` flag going to be correct? My gut tells me that a nested render cycle is going to screw with that global state.

Further, what "lifecycle method" is this safe in? `componentWillReceiveProps`?

jsfb replied:
@sebmarkbage said, for now, only componentDidMount and componentDidUpdate are allowed.  Currently calling it from a callback is not allowed (although I don't like this, I think it should be allowed, it's always easy to remove this restriction later).  Calling it twice in a row shouldn't be a problem, though I can't imagine why someone would want to do that.  Calling it from within Render is (according to sebastian) not good and disallowed.  I'll clarify the error message.

jsfb commented:
Returning value, updated error message to be more specific about when it's allowed.

Nested renders shouldn't be a problem since renders are not allowed to call this function (atm).  Although now that I think about it, @sebmarkbage is there a reason that render shouldn't be allowed?  Seems like it should be legal there, though it seems to me it should be legal everywhere.

zpao commented:
Let's drop the lifecycle stuff like we talked about.

jsfb commented:
Lifecycle stuff gone.

zpao commented on examples/basic-jsx/index.html:
```diff
@@ -40,12 +40,10 @@
       var start = new Date().getTime();
-      setInterval(function() {
-        React.render(
+        console.log(React.render(
```
Please revert

jsfb replied:
Done.

zpao commented:
There's a console.log that you missed. https://github.com/facebook/react/pull/3640/files#diff-c818825b9093d7a157d104790f972ff2R467

zpao commented on src/addons/__tests__/renderSubtreeIntoContainer.js:
```diff
@@ -0,0 +1,96 @@
+      componentDidMount: function() {
+        expect(function() {
+          renderSubtreeIntoContainer(<Parent />, <Component />, portal);
+        }.bind(this)).toThrow('Invariant Violation: parentComponent' +
```
Do you need to bind this?

jsfb replied:
Not in this case, that came from a copy-paste of the previous test.

zpao commented on src/core/ReactContext.js:
```diff
@@ -20,7 +20,6 @@ var emptyObject = require('emptyObject');
  * and is accessible via `this.context` on ReactCompositeComponents.
  */
 var ReactContext = {
-
```
Don't make this unneccessary change. If you were making any other changes in this file, fine. But you're not.

jsfb replied:
Introduced when I added the lifecycle stuff and then removed it.  Reverting the file.

zpao commented:
alright, lets do it.

jquense commented:
YA! 

Question. 

Is there a reason this is going into addons and not directly on react? It seems like this is a necessary function for anyone doing portals/layers/whatever its called, The switch to parent context will suddenly break anyone relying on this pattern, and force them to use react.addons when they didn't need to before. In particular this seems like a no-go for portal libraries, where the pattern in the community seems to be no use of addons, and if addons are used they either copy the code to their repo or `require()` the module directly from lib in order to avoid requiring consumers of the library to use react with addons, the use of both leading to duplication via webpack, etc.

zpao commented:
We're not sold on this API and we're working to see if we can make it go away entirely before we ship it. See Sebastian's comment - https://github.com/facebook/react/pull/3640#issuecomment-92533429.

The whole point of addons was that we could experiment more freely without the same levels of guarantees you get on the React object proper.
